### PR TITLE
BC-157 - Add typescript to task store

### PR DESCRIPTION
**Pull request by uidp**
# Short Description
BC-157 - Add typescript to task store

## Checklist before merging

- [x] QA: In addition to review, the code has been manually tested (if manual testing is possible)
- [x] PO: Any deviation from requirements was agreed with Product-Owner / ticket author / support-team

> Notice: Please keep this Pull-Request as a Draft (or add WIP label), until it is ready to be reviewed

### CODE REVIEW

File: src/store/tasks.ts [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+
+		return tasksCount;
+	}
+
+	private get isReady(): boolean {
+		return this.status === "completed";
+	}
+
+	private get taskApi() {
+		if (!this._taskApi) {
```
Comment (CeEv) [inline src/store/tasks.ts:336]: How the pagination is resolved in this context? (fetchAll)

File: src/store/tasks.unit.ts [hunk 2]
```diff
@@ -0,0 +1,806 @@
...
+		.spyOn(TaskFilter.prototype, method)
+		.mockReturnValue(new TaskFilter(result));
+}
+
+describe("task store", () => {
+	describe("actions", () => {
+		describe("getAllTasks", () => {
+			it("should request a list of tasks", (done) => {
+				const mockApi = {
+					taskControllerFindAll: jest.fn(() => ({
```
Comment (CeEv) [inline src/store/tasks.unit.ts:33]: It shoul work without pagination.

File: src/store/tasks.ts [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+	stateFactory: true,
+})
+export class TaskModule extends VuexModule {
+	tasks: Task[] = [];
+
+	courseFilter: string[] = [];
+
+	substituteFilter: boolean = false;
+
+	pagination: Pagination = {
```
Comment (CeEv) [inline src/store/tasks.ts:94]: We have no pagination that is stored at the place in context of not finished tasks.

File: src/store/tasks.unit.ts [hunk 4]
```diff
@@ -0,0 +1,806 @@
+import { Task, TaskModule } from "./tasks";
```
Comment (CeEv) [inline src/store/tasks.unit.ts:1]: I look into it after the requested changes are applayed. ( mark me self @CeEv )

**Review by CeEv — CHANGES_REQUESTED** [review]

File: src/store/task.filter.ts [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+		return new TaskFilter(filteredTasks);
+	}
+
+	withoutDueDate(): TaskFilter {
+		const withoutDueDate = this.tasks.filter((task) => !task.duedate);
+
+		return new TaskFilter(withoutDueDate);
+	}
+
+	withDueDate(): TaskFilter {
```
[the remark this answers is not in the document]
Comment (uidp) [reply src/store/task.filter.ts:77]: yes. we have to also think about how we can get rid of the `new Date()` comparison. the business logic shouldn't really depend on the client's current system time. that's maybe something we have to resolve soon.

File: src/store/task.filter.ts [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+
+	byOverdue(): TaskFilter {
+		const overdue = this.tasks.filter(
+			(task) => task.duedate && new Date(task.duedate) < new Date()
+		);
+
+		return new TaskFilter(overdue);
+	}
+
+	// it is a teacher based interpretation or and why or condition?
```
[the remark this answers is not in the document]
Comment (uidp) [reply src/store/task.filter.ts:93]: resolved by merge.

File: src/store/task.filter.unit.ts [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+		let list: Task[] = [];
+		for (let i = 0; i < number; i++) {
+			list.push(this.build(params));
+		}
+
+		return list;
+	}
+}
+
+export const taskFactory = new TaskFactory();
```
[the remark this answers is not in the document]
Comment (uidp) [reply src/store/task.filter.unit.ts:49]: yes. we should do that. but we first have to decide how we want to use factories and where to put the files. like we have done with the files under `store/types`. we maybe can use a path like `store/factories`

File: src/store/tasks.ts (hunk 3 above)
Comment (uidp) [reply src/store/tasks.ts:37]: right. removed that.

File: src/store/tasks.ts [hunk 8]
```diff
@@ -0,0 +1,347 @@
...
+			this.setTasks(response.data.data);
+			this.setStatus("completed");
+		} catch (error) {
+			this.setBusinessError(error as BusinessError);
+			this.setStatus("error");
+		}
+	}
+
+	@Mutation
+	setTasks(tasks: Task[]) {
```
[the remark this answers is not in the document]
Comment (uidp) [reply src/store/tasks.ts:67]: right. added that.

File: src/store/tasks.ts [hunk 9]
```diff
@@ -0,0 +1,347 @@
...
+	}
+
+	@Mutation
+	setCourseFilters(courseNames: string[]) {
+		this.courseFilter = courseNames;
+	}
+
+	@Mutation
+	setSubstituteFilter(enabled: boolean) {
+		this.substituteFilter = enabled;
```
[the remark this answers is not in the document]
Comment (uidp) [reply src/store/tasks.ts:78]: that has been discussed and i think it's resolved now.

File: src/store/tasks.ts [hunk 10]
```diff
@@ -0,0 +1,347 @@
...
+	};
+
+	businessError: BusinessError = {
+		statusCode: "",
+		message: "",
+	};
+
+	status: Status = "";
+
+	_taskApi?: TaskApiInterface;
```
[the remark this answers is not in the document]
Comment (uidp) [reply src/store/tasks.ts:50]: we have to access the task api somehow. this is the reference to it.

File: src/components/templates/TasksDashboardMain.vue [hunk 11]
```diff
@@ -163,15 +161,16 @@ export default {
 		},
 	},
 	mounted() {
-		this.$store.dispatch("tasks/getAllTasks");
+		TaskModule.getAllTasks();
```
[the remark this answers is not in the document]
Comment (uidp) [reply src/components/templates/TasksDashboardMain.vue:180]: yes. works the same in all components and pages.

File: src/components/templates/TasksDashboardMain.unit.js [hunk 12]
```diff
@@ -281,32 +132,76 @@ describe("@components/templates/TasksDashboardMain", () => {
...
-			},
-		});
-
-		expect(wrapper.vm.isFilterDisabled).toBe(true);
-		wrapper.setData({ tab: 1 });
-		expect(wrapper.vm.isFilterDisabled).toBe(false);
+		spy1.mockRestore();
 	});
 
+	// it("Should disable filter when active tab contains empty list and no course is selected", () => {
```
Comment (CeEv) [inline src/components/templates/TasksDashboardMain.unit.js:140]: Should this comment stand in?

File: src/components/templates/TasksDashboardMain.unit.js [hunk 13]
```diff
@@ -318,44 +213,53 @@ describe("@components/templates/TasksDashboardMain", () => {
...
-
-			expect(wrapper.vm.getSortedCoursesFilters[2]).toStrictEqual({
-				isSubstitution: false,
-				text: "Mathe (0)",
-				value: "Mathe",
-			});
-
-			expect(wrapper.vm.getSortedCoursesFilters).toHaveLength(3);
-		});
+		// it("Should set correct course name with task count for filter", () => {
```
Comment (CeEv) [inline src/components/templates/TasksDashboardMain.unit.js:216]: Should this comment stand in?

File: tests/unit/setup.js [hunk 14]
```diff
@@ -8,6 +8,13 @@ import path from "path";
 import commonTest from "./commonTests.js";
 import { RouterLinkStub } from "@vue/test-utils";
 
+// make object properties configurable in tests
+// so they can be mocked
+// https://stackoverflow.com/questions/58852549/error-spyonproperty-function-is-not-declared-configurable
+const { defineProperty } = Object;
+Object.defineProperty = (o, p, c) =>
```
Comment (CeEv) [inline tests/unit/setup.js:15]: This is a fine solution for tests.

File: src/store/task.filter.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (CeEv) [reply src/store/task.filter.ts:70]: Odalyse want to make a list of needed refacotor changes. We should add this also to it. @odalys-dataport;

**Review by odalys-dataport — APPROVED** [review]

**Pull request merged**